src/main/navigation/navigation_fixedwing.c

```diff
@@ -412,57 +412,38 @@ int16_t applyFixedWingMinSpeedController(timeUs_t currentTimeUs)
+        // ROLL >0 right, <0 left
+        int16_t rollCorrection = constrain(posControl.rcAdjustment[ROLL], -DEGREES_TO_DECIDEGREES(navConfig()->fw.max_bank_angle), DEGREES_TO_DECIDEGREES(navConfig()->fw.max_bank_angle));
+        rcCommand[ROLL] = pidAngleToRcCommand(rollCorrection, pidProfile()->max_angle_inclination[FD_ROLL]);
     }
 
     if (isPitchAdjustmentValid && (navStateFlags & NAV_CTL_ALT)) {
-        pitchCorrection += posControl.rcAdjustment[PITCH];
-        throttleCorrection += DECIDEGREES_TO_DEGREES(pitchCorrection) * navConfig()->fw.pitch_to_throttle;
+        // PITCH >0 dive, <0 climb
+        int16_t pitchCorrection = constrain(posControl.rcAdjustment[PITCH], -DEGREES_TO_DECIDEGREES(navConfig()->fw.max_dive_angle), DEGREES_TO_DECIDEGREES(navConfig()->fw.max_climb_angle));
+        rcCommand[PITCH] = -pidAngleToRcCommand(pitchCorrection, pidProfile()->max_angle_inclination[FD_PITCH]);
+        int16_t throttleCorrection = DECIDEGREES_TO_DEGREES(pitchCorrection) * navConfig()->fw.pitch_to_throttle;
 
 #ifdef NAV_FIXED_WING_LANDING
         if (navStateFlags & NAV_CTL_LAND) {
-            /*
-             * During LAND we do not allow to raise THROTTLE when nose is up
-             * to reduce speed
-             */
+            // During LAND we do not allow to raise THROTTLE when nose is up to reduce speed
             throttleCorrection = constrain(throttleCorrection, minThrottleCorrection, 0);
         } else {
 #endif
             throttleCorrection = constrain(throttleCorrection, minThrottleCorrection, maxThrottleCorrection);
 #ifdef NAV_FIXED_WING_LANDING
         }
 #endif
-    }
-
-    // Speed controller - only apply in POS mode when NOT NAV_CTL_LAND
-    if ((navStateFlags & NAV_CTL_POS) && !(navStateFlags & NAV_CTL_LAND)) {
-        throttleCorrection += applyFixedWingMinSpeedController(currentTimeUs);
-        throttleCorrection = constrain(throttleCorrection, minThrottleCorrection, maxThrottleCorrection);
-    }
-
-    // Limit and apply
-    if (isPitchAdjustmentValid && (navStateFlags & NAV_CTL_ALT)) {
-        // PITCH correction is measured according to altitude: <0 - dive/lose altitude, >0 - climb/gain altitude
-        // PITCH angle is measured in opposite direction ( >0 - dive, <0 - climb)
-        pitchCorrection = constrain(pitchCorrection, -DEGREES_TO_DECIDEGREES(navConfig()->fw.max_dive_angle), DEGREES_TO_DECIDEGREES(navConfig()->fw.max_climb_angle));
-        rcCommand[PITCH] = -pidAngleToRcCommand(pitchCorrection, pidProfile()->max_angle_inclination[FD_PITCH]);
-    }
 
-    if (isRollAdjustmentValid && (navStateFlags & NAV_CTL_POS)) {
-        rollCorrection = constrain(rollCorrection, -DEGREES_TO_DECIDEGREES(navConfig()->fw.max_bank_angle), DEGREES_TO_DECIDEGREES(navConfig()->fw.max_bank_angle));
-        rcCommand[ROLL] = pidAngleToRcCommand(rollCorrection, pidProfile()->max_angle_inclination[FD_ROLL]);
-    }
+        // Speed controller - only apply in POS mode when NOT NAV_CTL_LAND
```

Comment:
Now this is applied only if both ALT and POS controllers are active, while previously speed guard only required POS which is reasonable. Why change?

Reply:
For the moment all nav states using POS also use ALT so it won't change anything for those but for cruise 2D we need manual throttle control (no ALT) and we can't have both manual throttle control and min speed applied.

Reply:
Ok, you convinced me :smile: